# Project: Wrangle WeRateDogs data from Twitter

Table of Contents
1. Introduction
2. Data Wrangling
3. Storing and Acting on Wrangling Data

1.Introduction
This project is about wrangling data for the Wrangle and Analyze Project
The document is divided in 3 sections, each for one step of the dta wrangling process that we saw in the lessons

Context
Our goal: wrangle WeRateDogs Twitter data to create interesting and trustworthy analyses and visualizations. The Twitter archive is great, but it only contains very basic tweet information. Additional gathering, then assessing and cleaning is required for "Wow!"-worthy analyses and visualizations.

Project Details
Our tasks in this project are as follows:

Data wrangling, which consists of:
Gathering data (downloadable file in the Resources tab in the left most panel of your classroom and linked in step 1 below).
Assessing data
Cleaning data
Storing, analyzing, and visualizing your wrangled data
Reporting on 1) your data wrangling efforts and 2) your data analyses and visualizations


2. Data Wrangling

Gather
First we need to gather the data from at least 3 different sources. Some data were wprovides and other were obtained

In [5]:
#first load all the libraries needed

import pandas as pd
import numpy as np
import requests
import json
import tweepy
import os
import time
import re
import matplotlib.pyplot as plt
import warnings
from IPython.display import Image
from functools import reduce
import re
import seaborn as sns
import datetime
%matplotlib inline

In [7]:
#Now load the twitter_archive_enhanced.csv file into the dataframe
tweet_data = pd.read_csv("twitter-archive-enhanced.csv")
tweet_data.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [8]:
# Now, we retrieve image-predictions.tsv file
prediction_url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
with open('image-predictions.tsv', 'wb') as prediction_file:
    prediction_gather = requests.get(prediction_url)
    prediction_file.write(prediction_gather.content)
    
#Now we retrieve the image_predictions.tsv file into its dataframe
prediction_data = pd.read_csv("image_predictions.tsv",sep='\t')
prediction_data.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [11]:
# Now we load the twtter data file
consumer_key = 'XXX'
consumer_secret = 'XXX'
access_token = 'XXX'
access_secret = 'XXX'
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)
api = tweepy.API(auth)


In [20]:
tweet_json = pd.read_json("tweet-json.txt", lines=True)

tweet_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   created_at                     2354 non-null   datetime64[ns, UTC]
 1   id                             2354 non-null   int64              
 2   id_str                         2354 non-null   int64              
 3   full_text                      2354 non-null   object             
 4   truncated                      2354 non-null   bool               
 5   display_text_range             2354 non-null   object             
 6   entities                       2354 non-null   object             
 7   extended_entities              2073 non-null   object             
 8   source                         2354 non-null   object             
 9   in_reply_to_status_id          78 non-null     float64            
 10  in_reply_to_status_id_st

Assess

Assessing our data is the second step in data wrangling. When assessing, we are like a detective at work, inspecting our dataset for two things: data quality issues (i.e. content issues) and lack of tidiness (i.e. structural issues).

Quality issues are issues with content, like inaccurate or duplicate data. Tidiness issues are structural issues, specifically: each variable must be a column, each observation must be a row, and each type of observational unit must be a table.

We will use two steps: Visual Assessment and Programmatic Assessment

1. Dataframe contents for visual assessment

In [21]:
# Display the twitter data dataframe
tweet_data

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

Visual
Quality Issues
1. Name column, have "None" as name. This can be confusing
2. There is missing data in "name" columns and is written as the string "None"

Tidity Issues
1. The columns "doggo", "floofer", "pupper" and "puppo" represent the same variable "type"

In [22]:
# Display the prediction data dataframe
prediction_data

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

Visual Quality Issues

3. In p1, p2, and p3 the type of dogs uses underscores that are diffucult to read

Tidity Issues

2. Correct predictions of dogs should be combined with tweet_data

2. Dataframe contents for programmatic assessment

In [23]:
#We display the structure of the dataframe tweet_data
tweet_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

Quality Issues
4. Missing "expanded URLS", there are 2297 instead of 2356
5. retweeted_status_id need to be "int64"
6. timestamp is better to be "date_time" objet type
7. there are retweet data, it shouldn´t

In [24]:
#look for more issues with sample rows
tweet_data.sample(15)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1359  703356393781329922                    NaN                  NaN   
38    884925521741709313                    NaN                  NaN   
744   780192070812196864                    NaN                  NaN   
1757  678764513869611008                    NaN                  NaN   
1660  683078886620553216                    NaN                  NaN   
461   817536400337801217                    NaN                  NaN   
1173  720340705894408192                    NaN                  NaN   
929   754482103782404096                    NaN                  NaN   
62    880095782870896641                    NaN                  NaN   
705   785639753186217984                    NaN                  NaN   
1146  727155742655025152                    NaN                  NaN   
1157  724405726123311104                    NaN                  NaN   
2291  667165590075940865                    NaN                  NaN   
716   783821107061198850                    NaN                  NaN   
2318  666454714377183233                    NaN                  NaN   

                      timestamp  \
1359  2016-02-26 23:10:06 +0000   
38    2017-07-12 00:01:00 +0000   
744   2016-09-25 23:47:39 +0000   
1757  2015-12-21 02:30:45 +0000   
1660  2016-01-02 00:14:32 +0000   
461   2017-01-07 01:00:41 +0000   
1173  2016-04-13 19:59:42 +0000   
929   2016-07-17 01:05:25 +0000   
62    2017-06-28 16:09:20 +0000   
705   2016-10-11 00:34:48 +0000   
1146  2016-05-02 15:20:13 +0000   
1157  2016-04-25 01:12:38 +0000   
2291  2015-11-19 02:20:46 +0000   
716   2016-10-06 00:08:09 +0000   
2318  2015-11-17 03:16:00 +0000   

                                                 source  \
1359  <a href="http://twitter.com/download/iphone" r...   
38    <a href="http://twitter.com/download/iphone" r...   
744   <a href="http://twitter.com/download/iphone" r...   
1757  <a href="http://twitter.com/download/iphone" r...   
1660  <a href="http://twitter.com/download/iphone" r...   
461   <a href="http://twitter.com/download/iphone" r...   
1173  <a href="http://twitter.com/download/iphone" r...   
929   <a href="http://twitter.com/download/iphone" r...   
62    <a href="http://twitter.com/download/iphone" r...   
705   <a href="http://twitter.com/download/iphone" r...   
1146  <a href="http://vine.co" rel="nofollow">Vine -...   
1157  <a href="http://twitter.com/download/iphone" r...   
2291  <a href="http://twitter.com/download/iphone" r...   
716   <a href="http://twitter.com/download/iphone" r...   
2318  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1359  This is Socks. That water pup w the super legs...                  NaN   
38    This is Earl. He found a hat. Nervous about wh...                  NaN   
744   We only rate dogs. Pls stop sending in non-can...                  NaN   
1757  Meet Wilson. He got caught humping the futon. ...                  NaN   
1660  Here we see a nifty leaping pupper. Feet look ...                  NaN   
461   Say hello to Eugene &amp; Patti Melt. No matte...                  NaN   
1173  This is Derek. He just got balled on. Can't ev...                  NaN   
929   This is Milo. He's currently plotting his reve...                  NaN   
62    Please don't send in photos without dogs in th...                  NaN   
705   This is Pinot. He's a sophisticated doggo. You...                  NaN   
1146  When you're trying to enjoy yourself but end u...                  NaN   
1157  This is Ashleigh. She's having Coachella withd...                  NaN   
2291  This is Churlie. AKA Fetty Woof. Lost eye savi...                  NaN   
716   This is Mairi. She has mastered the art of cam...                  NaN   
2318  I'll name the dogs from now on. This is Kreggo...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1359           

In [25]:
# Look for quality issues in the range of data
tweet_data[['rating_numerator', 'rating_denominator']].describe()

rating_numerator  rating_denominator
count       2356.000000         2356.000000
mean          13.126486           10.455433
std           45.876648            6.745237
min            0.000000            0.000000
25%           10.000000           10.000000
50%           11.000000           10.000000
75%           12.000000           10.000000
max         1776.000000          170.000000

Quality Issues
8. There is a minimum value of 0 in the rating_denominator column
9. The maximun value is higher for both numerator and denominator

In [26]:
#Get the value of the numerator column
tweet_data.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
0         2
15        2
75        2
420       2
182       1
204       1
143       1
121       1
99        1
20        1
45        1
27        1
17        1
24        1
26        1
44        1
50        1
60        1
80        1
84        1
88        1
1776      1
960       1
666       1
144       1
165       1
Name: rating_numerator, dtype: int64

In [27]:
#Dig more for the denominator equals zero
tweet_data.rating_denominator.value_counts()

10     2333
11        3
50        3
20        2
80        2
0         1
120       1
7         1
170       1
150       1
130       1
90        1
110       1
2         1
70        1
40        1
16        1
15        1
Name: rating_denominator, dtype: int64

There is only 1 case that the denominator is "0"

In [28]:
#Looking for duplicated tweet_id
tweet_data.tweet_id.value_counts()

700151421916807169    1
674764817387900928    1
794926597468000259    1
701570477911896070    1
715220193576927233    1
                     ..
674742531037511680    1
756651752796094464    1
743545585370791937    1
770655142660169732    1
749075273010798592    1
Name: tweet_id, Length: 2356, dtype: int64

No duplicate tweet_id

In [29]:
# Display the structure of prediction data dataframe
prediction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


Quality Issues
10. There are 2075 entries and in the tweet_data dataframe there are 2376 rows, so there is missing data

In [30]:
#look for duplicate data in prediction_data dataframe
prediction_data.tweet_id.value_counts()

700151421916807169    1
740373189193256964    1
836989968035819520    1
816336735214911488    1
688828561667567616    1
                     ..
675354435921575936    1
759923798737051648    1
705475953783398401    1
693231807727280129    1
817056546584727552    1
Name: tweet_id, Length: 2075, dtype: int64

No duplicate data

In [31]:
# Compute statistics on the p*_conf columns in prediction_data
prediction_data[['img_num', 'p1_conf', 'p2_conf', 'p3_conf']].describe()

img_num      p1_conf       p2_conf       p3_conf
count  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean      1.203855     0.594548  1.345886e-01  6.032417e-02
std       0.561875     0.271174  1.006657e-01  5.090593e-02
min       1.000000     0.044333  1.011300e-08  1.740170e-10
25%       1.000000     0.364412  5.388625e-02  1.622240e-02
50%       1.000000     0.588230  1.181810e-01  4.944380e-02
75%       1.000000     0.843855  1.955655e-01  9.180755e-02
max       4.000000     1.000000  4.880140e-01  2.734190e-01

No quality Issues, no negative numbers, , range between [0,1]

Summary of Quality and Tidiness Issues

Quality
1. Name column, have "None" as name. This can be confusing
2. There is missing data in "name" columns and is written as the string "None"
4. Missing "expanded URLS", there are 2297 instead of 2356
5. retweeted_status_id need to be "int64"
6. timestamp is better to be "date_time" objet type
7. there are retweet data, it shouldn´t
8. There is a minimum value of 0 in the rating_denominator column
9. The maximun value is higher for both numerator and denominator
3. In p1, p2, and p3 the type of dogs uses underscores that are diffucult to read
10. There are 2075 entries and in the tweet_data dataframe there are 2376 rows, so there is missing data

Tidiness
1. The columns "doggo", "floofer", "pupper" and "puppo" represent the same variable "type"
2. Correct predictions of dogs should be combined with tweet_data

Clean

Cleaning our data is the third step in data wrangling. It is where we fix the quality and tidiness issues that we identified in the assess step

We will use the data cleaning process: defining, coding, and testing
•	We will address the missing data first
•	We will tackle the tidiness issues next
•	And finally, we will clean up the quality issues

The very first thing to do before any cleaning occurs is to make a copy of each piece of data. All of the cleaning operations will be conducted on this copy so we can still view the original dirty and/or messy dataset later

In [33]:
#create a copy of each dataframe and merge
tweet_data_clean = tweet_data.copy()
prediction_data_clean = prediction_data.copy()
tweet_json_clean = tweet_json.copy()

# Ref: https://stackoverflow.com/questions/44327999/python-pandas-merge-multiple-dataframes/44338256
dfs = pd.concat([tweet_data_clean, prediction_data_clean, tweet_json_clean], join='outer', axis=1)
dfs.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  ...  favorite_count  \
0  https://twitter.com/dog_rates/status/892420643...  ...         39467.0   
1  https://twitter.com/dog_rates/status/892177421...  ...         33819.0   
2  https://twitter.com/dog_rates/status/891815181...  ...         25461.0   
3  https://twitter.com/dog_rates/status/891689557...  ...         42908.0   
4  https://twitter.com/dog_rates/status/891327558...  ...         41048.0   

   favorited retweeted possibly_sensitive possibly_sensitive_appealable lang  \
0      False     False                0.0                           0.0   en   
1      False     False                0.0                           0.0   en   
2      False     False                0.0                           0.0   en   
3      False     False                0.0                           0.0   en   
4      False     False                0.0                           0.0   en   

  retweeted_status  quoted_status_id quoted_status_id_str  quoted_status  
0              NaN               NaN                  NaN            NaN  
1              NaN               NaN                  NaN            NaN  
2              NaN               NaN                  NaN            NaN  
3              NaN               NaN                  NaN            NaN  
4              NaN               NaN                  NaN            NaN  

[5 rows x 60 columns]

Quality

6. timestamp is better to be "date_time" objet type

Define
Convert the timestamp column into datetime

Code

In [34]:
#Changetimestamp column to datetime type
ts_col_names = [t_timestamp for t_timestamp in tweet_data_clean.columns.tolist() if 'timestamp' in t_timestamp]
for ts_col_name in ts_col_names:
    tweet_data_clean[ts_col_name] = pd.to_datetime(tweet_data_clean[ts_col_name])

Test

In [35]:
tweet_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2356 non-null   int64              
 1   in_reply_to_status_id       78 non-null     float64            
 2   in_reply_to_user_id         78 non-null     float64            
 3   timestamp                   2356 non-null   datetime64[ns, UTC]
 4   source                      2356 non-null   object             
 5   text                        2356 non-null   object             
 6   retweeted_status_id         181 non-null    float64            
 7   retweeted_status_user_id    181 non-null    float64            
 8   retweeted_status_timestamp  181 non-null    datetime64[ns, UTC]
 9   expanded_urls               2297 non-null   object             
 10  rating_numerator            2356 non-null   int64           

7. there are retweet data, it shouldn´t

Define
Remove all rows where retweeted_status_id. Afterwards, reset the index.

Code

In [36]:
retweeted_cols = ['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp']
retweeted_mask = tweet_data_clean.retweeted_status_id.notnull()

tweet_data_clean = tweet_data_clean[~retweeted_mask]
tweet_data_retweet_test = tweet_data_clean.copy() # For testing
tweet_data_clean.drop(columns=retweeted_cols, inplace=True)
tweet_data_clean.reset_index(drop=True, inplace=True)

Test

In [37]:
# check if we deleted all of the retweets
tweet_data_retweet_test.retweeted_status_id.notnull().sum()

0

1. Name column, have "None" as name. This can be confusing

Define
We will replace the string "None" with "NaN"

Code

In [38]:
tweet_data_clean.name = tweet_data_clean.name.replace('None', np.nan)

test

In [39]:
tweet_data_clean.name.loc[tweet_data_clean.name == 'None'].sum()

0

In [40]:
# Tidy issue # 1
# Code and test
pd.melt(
    tweet_data_clean, 
    id_vars = ['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp', 'source','text','expanded_urls', 'rating_numerator','rating_denominator','name'],
    var_name ='stages',
    value_name = 'val')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
8695  666049248165822465                    NaN                  NaN   
8696  666044226329800704                    NaN                  NaN   
8697  666033412701032449                    NaN                  NaN   
8698  666029285002620928                    NaN                  NaN   
8699  666020888022790149                    NaN                  NaN   

                     timestamp  \
0    2017-08-01 16:23:56+00:00   
1    2017-08-01 00:17:27+00:00   
2    2017-07-31 00:18:03+00:00   
3    2017-07-30 15:58:51+00:00   
4    2017-07-29 16:00:24+00:00   
...                        ...   
8695 2015-11-16 00:24:50+00:00   
8696 2015-11-16 00:04:52+00:00   
8697 2015-11-15 23:21:54+00:00   
8698 2015-11-15 23:05:30+00:00   
8699 2015-11-15 22:32:08+00:00   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
8695  <a href="http://twitter.com/download/iphone" r...   
8696  <a href="http://twitter.com/download/iphone" r...   
8697  <a href="http://twitter.com/download/iphone" r...   
8698  <a href="http://twitter.com/download/iphone" r...   
8699  <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
0     This is Phineas. He's a mystical boy. Only eve...   
1     This is Tilly. She's just checking pup on you....   
2     This is Archie. He is a rare Norwegian Pouncin...   
3     This is Darla. She commenced a snooze mid meal...   
4     This is Franklin. He would like you to stop ca...   
...                                                 ...   
8695  Here we have a 1949 1st generation vulpix. Enj...   
8696  This is a purebred Piers Morgan. Loves to Netf...   
8697  Here is a very happy pup. Big fan of well-main...   
8698  This is a western brown Mitsubishi terrier. Up...   
8699  Here we have a Japanese Irish Setter. Lost eye...   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitter.com/dog_rates/status/891689557...                13   
4     https://twitter.com/dog_rates/status/891327558...                12   
...                                                 ...               ...   
8695  https://twitter.com/dog_rates/status/666049248...                 5   
8696  https://twitter.com/dog_rates/status/666044226...                 6   
8697  https://twitter.com/dog_rates/status/666033412...                 9   
8698  https://twitter.com/dog_rates/status/666029285...                 7   
8699  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator      name stages   val  
0                     10   Phineas  doggo  None  
1                     10     Tilly  doggo  None  
2                     10    Archie  doggo  None  
3                     10     Darla  doggo  None  
4                     10  Franklin  doggo  None  
...                  ...       ...    ...   ...  
8695                  10       NaN  puppo  None  

3. Storing and Acting on Wrangling Data

Storing

In [41]:
# save our dataframe in the file specified
tweet_data_clean.to_csv("twitter_archive_master.csv")

In [42]:
# save our dataframe in the file specified
prediction_data_clean.to_csv("twitter_archive_secondary.csv")

Work in progress

Analysis and Insights
3 insights and 1 labeled visualization

In [43]:
#Most common scores on the tweets
tweet_data_clean['score'] = tweet_data_clean['rating_numerator'].astype(str) + "/" + \
                                  tweet_data_clean['rating_denominator'].astype(str)
tweet_data_clean.score.value_counts()

12/10      500
10/10      442
11/10      425
13/10      307
9/10       155
8/10        98
7/10        53
14/10       43
5/10        36
6/10        32
3/10        19
4/10        15
2/10         9
1/10         7
4/20         2
0/10         2
420/10       2
99/90        1
75/10        1
88/80        1
26/10        1
204/170      1
144/120      1
143/130      1
1776/10      1
60/50        1
9/11         1
84/70        1
960/0        1
44/40        1
20/16        1
50/50        1
121/110      1
1/2          1
80/80        1
45/50        1
17/10        1
182/10       1
24/7         1
11/15        1
7/11         1
27/10        1
666/10       1
165/150      1
15/10        1
Name: score, dtype: int64

Insight:
The most common scores are around 10/10 +1 or -1